### DISCUSSION TITLE
Autopilot - Get to a point with correct angle

### DISCUSSION THREAD

**Discussion by lauritowal** [JSBSim-Team/jsbsim#410] [Q&A]
Hi!
Is it possible to use an autopilot in jsbsim to get an aircraft (e.g. Cesna) from point A (inital position) to point B (e.g. position 
just before starting with landing procedure). Furthermore, when the aircraft arrives at point B I would like the aircraft to have a specific heading angle (e.g. 30°). 

Is it possible doing this with autopilots provided by jsbsim and if yes how would I do that in Python? I've attached also an image if that helps clarify what I mean. 
Right now, I have created a custom heading hold script based on a chapter in  [Priciples of Flight Simulations](https://onlinelibrary.wiley.com/doi/10.1002/9780470685662.ch4) but there I am not sure how to get the aircraft to be at a specific angle when it does arrive at point B. That's why I was wondering if in JSBsim this is already implemented somehow...

Thanks for all your help and advice in advance :)
![Screenshot 2021-04-08 at 11 52 08](https://user-images.githubusercontent.com/3471924/114006606-e0290880-9860-11eb-8165-9142c5677b66.png)

**Comment by seanmcleod** [comment]
Hmm, Reinforcement Learning and JSBSim seems be very popular, majority of the questions we get recently have been from people using JSBSim for RL.

I'm not aware of any of the autopilot examples in the JSBSim repo providing exactly what you want, they're mostly more basic demonstrations of how to go about adding basic autopilot functionality.

You may find that some of the FlightGear aircraft implement more fully featured autopilots, although in doing so they may make use of LUA scripting to help them implement the required functionality. Which isn't supported by stand-alone JSBSim.

In effect you're looking for some sort of trajectory planning and following autopilot, and for your particular request you're looking to arrive at the waypoint "on track" for the following waypoint (runway threshold in your case).

So a turn needs to be initiated before reaching point B (Current waypoint in my example) and the size of the arc of the turn will depend on the TAS and the amount of g to be used in the turn. 

![image](https://user-images.githubusercontent.com/824856/114012477-a5c26a00-9866-11eb-9c66-3c27e516de78.png)

Here is another example, in this case there is also optionally a track range (FDTK) of some specified distance that the aircraft needs to attain the next track by, could be 0. As I mentioned the circle's radius is dependent on the TAS and g you want to use.

![image](https://user-images.githubusercontent.com/824856/114013015-5df01280-9867-11eb-9841-8540a337e436.png)

**Reply by lauritowal** [reply]
Thank you very much for your answer @seanmcleod! 
Is the method you've mentioned "Turn ancitipation"? Just to know what I have to google for :D I found that in the book which I've linked above and try to make sense from it.
Could you maybe provide a simple formula or pseudo code snippet to calculate the desired heading for each time step to reach the 'current waypoint / point B' with the correct heading for the method you've mentioned? Or point me towards the right ressources to understand the whole idea better and the images you've added?

Thanks again for your help!
